File: src/main/java/com/addonovan/cse3330/DatabaseDriver.java
package com.addonovan.cse3330;

import com.addonovan.cse3330.model.Profile;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public final class DatabaseDriver {

    /** A URL used to open a connection to the SocialMedia database. */
    private static final String CONNECTION_STRING = "jdbc:postgresql://localhost/SocialMedia";

    /** The actual database connection. */
    private static final Connection CONNECTION;

    static {
        Properties props = new Properties();
        props.setProperty("user", "application");
        props.setProperty("password", "password1"); // oh boy! check that security!

        // if there's a problem connecting to the database, then we'll immediately
        // throw an exception and die off
        try {
            CONNECTION = DriverManager.getConnection(CONNECTION_STRING, props);
        } catch (SQLException e) {
            throw new RuntimeException("Failed to open database connection!", e);
        }

        // make sure the connection is (properly) closed on shutdown, not that
        // it *really* matters all that much
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            try {
                CONNECTION.close();
            } catch (SQLException e) {
                throw new RuntimeException("Failed to close database connection!", e);
            }
        }));
    }

    /**
     * @return A list of all active profiles.
     */
    public static List<Profile> listProfiles() {
        List<Profile> profiles = new ArrayList<>();

        try (Statement stmt = CONNECTION.createStatement()) {
            ResultSet results = stmt.executeQuery("SELECT * FROM Profile;");

            while (results.next()) {
                boolean active = results.getBoolean("ProfileActive");
                if (!active) continue;

                profiles.add(new Profile(
                        results.getInt("ProfileId"),
                        results.getString("FirstName"),
                        results.getString("LastName"),
                        results.getString("PhoneNumber"),
                        results.getString("Email"),
                        results.getString("UserName"),
                        results.getString("Password"),
                        results.getDate("CreatedTime"),
                        true
                ));
            }

        } catch (SQLException e) {
            throw new RuntimeException("Encountered error running query", e);
        }

        return profiles;
    }

}

File: src/main/java/com/addonovan/cse3330/DatabaseDriver.kt
package com.addonovan.cse3330

import com.addonovan.cse3330.model.Account
import com.addonovan.cse3330.model.Profile
import org.intellij.lang.annotations.Language

import java.sql.*
import java.util.Properties

object DatabaseDriver {

    /** A URL used to open a connection to the SocialMedia database.  */
    private val CONNECTION_STRING = "jdbc:postgresql://localhost/SocialMedia"

    /** The actual database connection.  */
    private val CONNECTION: Connection

    init {
        val props = Properties()
        props.setProperty("user", "application")
        props.setProperty("password", "password1") // oh boy! check that security!

        // if there's a problem connecting to the database, then we'll immediately
        // throw an exception and die off
        try {
            CONNECTION = DriverManager.getConnection(CONNECTION_STRING, props)
        } catch (e: SQLException) {
            throw RuntimeException("Failed to open database connection!", e)
        }

        // make sure the connection is (properly) closed on shutdown, not that
        // it *really* matters all that much
        Runtime.getRuntime().addShutdownHook(Thread {
            try {
                CONNECTION.close()
            } catch (e: SQLException) {
                throw RuntimeException("Failed to close database connection!", e)
            }
        })
    }

    private inline fun <T> createStatement(crossinline action: (Statement) -> T): T {
        try {
            CONNECTION.createStatement().use {
                return action(it)
            }
        }
        catch (e: SQLException) {
            throw RuntimeException("Failed to execute query!", e)
        }
    }

    private inline fun <T> query(query: String, crossinline action: (ResultSet) -> T): List<T> {
        val list = arrayListOf<T>()
        createStatement {
            val resultSet = it.executeQuery(query)
            while (resultSet.next()) {
                list += action(resultSet)
            }
        }
        return list
    }

    private fun insert(@Language("PostgreSQL") query: String): ResultSet {
        var set: ResultSet? = null
        createStatement {
            set = it.executeQuery(query)
        }
        return set ?: throw RuntimeException("INSERT failed: $query")
    }

    fun listAccounts() = query("""SELECT * FROM "Account" WHERE IsActive = TRUE;""") {
        Account.fromRow(it)
    }

    fun getProfileById(id: Int): Profile? {
        val profiles = query("""
            |SELECT *
            |FROM "Profile" p
            |INNER JOIN "Account" a
            |ON a.Id = p.AccountId
            |WHERE Id = $id
            """.trimMargin()) {
            Profile.fromRow(it)
        }
        return profiles.firstOrNull()
    }

    fun getAccountById(id: Int) = query("""SELECT * FROM "Account" WHERE Id = $id""") {
        Account.fromRow(it)
    }

    fun insertProfile(profile: Profile): Profile {
        val newAccount = profile.account.copy(id = insert(profile.account.insert()).getInt(0))
        val newProfile = profile.copy(account = newAccount)
        insert(newProfile.insert())
        return newProfile
    }

}
